stability: crash on trace

On a running cluster:
oddly, I only have one stack trace.

@tschottdorf: you're probably the most familiar with tracing code.
The relevant line is:
```
		// Discard the middle events.
		di := int((cap(tr.events) - 1) / 2)
		if d, ok := tr.events[di].What.(*discarded); ok {
```

```
W160302 20:57:40.694712 storage/replica.go:1745  [node=3,store=3,range=1] failed to resolve intents: range 1: replica node_id:3 store_id:3 replica_id:14  not leader; leader is node_id:2 store_id:2 replica_id:13
I160302 20:57:40.701867 storage/replica_command.go:1272  range 1: new leader lease replica {3 3 14} 2016-03-02 20:57:40 +0000 UTC +1.003s
W160302 20:57:40.702697 storage/replica.go:1745  [node=3,store=3,range=1] failed to resolve intents: context deadline exceeded
panic: runtime error: index out of range

goroutine 85 [running]:
golang.org/x/net/trace.(*trace).addEvent(0xc820a46600, 0x15ff280, 0xc827a97f20, 0xc837c30000)
        /go/src/golang.org/x/net/trace/trace.go:757 +0x92b
golang.org/x/net/trace.(*trace).LazyPrintf(0xc820a46600, 0x17ef700, 0xe, 0x0, 0x0, 0x0)
        /go/src/golang.org/x/net/trace/trace.go:786 +0xa2
github.com/cockroachdb/cockroach/util/tracing.(*netTraceWrapSpan).LogEvent(0xc836804180, 0x17ef700, 0xe)
        /go/src/github.com/cockroachdb/cockroach/util/tracing/nettrace.go:44 +0x67
github.com/cockroachdb/cockroach/storage.(*Replica).processRaftCommand(0xc8363061a0, 0xc84f9e5d60, 0x8, 0xc2ec, 0x1, 0x300000003, 0xe, 0x1438225cb06309ed, 0x1, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1265 +0x2a8
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc8363061a0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1130 +0xa53
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
        /go/src/github.com/cockroachdb/cockroach/storage/store.go:1762 +0x318
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202b5810, 0xc820153480)
        /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
        /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x62
```

still happening, somewhat regularly (eg: a few times a day on a 4 node cluster with load):
```
W160304 20:30:35.183289 storage/replica.go:1795  could not GC completed transaction: context deadline exceeded
W160304 20:30:35.184098 storage/replica.go:1751  [node=2,store=2,range=1] failed to resolve intents: context deadline exceeded
W160304 20:30:35.184225 storage/replica.go:1751  [node=2,store=2,range=1] failed to resolve intents: context deadline exceeded
I160304 20:30:35.184720 storage/replica_command.go:1272  range 33: new leader lease replica {2 2 8} 2016-03-04 20:30:29 +0000 UTC +6.922s
W160304 20:30:35.198180 storage/replica.go:1751  [node=2,store=2,range=1] failed to resolve intents: context deadline exceeded
panic: runtime error: index out of range

goroutine 64 [running]:
golang.org/x/net/trace.(*trace).addEvent(0xc821fb3680, 0x16074a0, 0xc8233f7da0, 0x1440000)
	/go/src/golang.org/x/net/trace/trace.go:757 +0x92b
golang.org/x/net/trace.(*trace).LazyPrintf(0xc821fb3680, 0x17dafb0, 0x2, 0xc824e8fbe0, 0x1, 0x1)
	/go/src/golang.org/x/net/trace/trace.go:786 +0xa2
github.com/opentracing/basictracer-go.glob.func1.1(0x1696e40, 0xc821e8dc00)
	/go/src/github.com/opentracing/basictracer-go/event_nettrace.go:19 +0x343
github.com/opentracing/basictracer-go.(*spanImpl).onLog(0xc84d9d5e60, 0x0, 0x0, 0x0, 0x17f9a70, 0xe, 0x0, 0x0)
	/go/src/github.com/opentracing/basictracer-go/event.go:40 +0xc4
github.com/opentracing/basictracer-go.(*spanImpl).Log(0xc84d9d5e60, 0xece6be76b, 0xc070c2c, 0x25b2b60, 0x17f9a70, 0xe, 0x0, 0x0)
	/go/src/github.com/opentracing/basictracer-go/span.go:97 +0x1e6
github.com/opentracing/basictracer-go.(*spanImpl).LogEvent(0xc84d9d5e60, 0x17f9a70, 0xe)
	/go/src/github.com/opentracing/basictracer-go/span.go:74 +0x7d
github.com/cockroachdb/cockroach/storage.(*Replica).processRaftCommand(0xc8252aef70, 0xc824e8fbb0, 0x8, 0x9bc4, 0x1, 0x200000002, 0x7, 0x1438be0b4c0ab6a1, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:1268 +0x2a8
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc8252aef70, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:1133 +0xa53
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1762 +0x318
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202f6a80, 0xc8202f32d0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x62
```

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -135,6 +135,7 @@ func usesTimestampCache(r roachpb.Request) bool {
 // via the done channel.
 type pendingCmd struct {
 	ctx     context.Context
+	idKey   cmdIDKey
 	raftCmd roachpb.RaftCommand
 	done    chan roachpb.ResponseWithError // Used to signal waiting RPC handler
 }
@@ -635,7 +636,6 @@ func (r *Replica) RaftStatus() *raft.Status {
 // range's leadership is confirmed. The command is then dispatched
 // either along the read-only execution path or the read-write Raft
 // command queue.
-// TODO(tschottdorf): use BatchRequest w/o pointer receiver.
 func (r *Replica) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.BatchResponse, *roachpb.Error) {
 	var br *roachpb.BatchResponse
 
@@ -963,16 +963,49 @@ func (r *Replica) addWriteCmd(ctx context.Context, ba roachpb.BatchRequest, wg *
 
 	if err == nil {
 		// If the command was accepted by raft, wait for the range to apply it.
-		select {
-		case respWithErr := <-pendingCmd.done:
-			br, pErr = respWithErr.Reply, respWithErr.Err
-		case <-ctx.Done():
-			pErr = roachpb.NewError(ctx.Err())
+		ctxDone := ctx.Done()
+		for br == nil && pErr == nil {
+			select {
+			case respWithErr := <-pendingCmd.done:
+				br, pErr = respWithErr.Reply, respWithErr.Err
+			case <-ctxDone:
+				// Cancellation is somewhat tricky since we can't prevent the
+				// Raft command from executing at some point in the future.
+				// We try to remove the pending command, but if the processRaft
+				// goroutine has already grabbed it (as would typically be the
+				// case right as it executes), it's too late and we're still
+				// going to have to wait until the command returns (which would
+				// typically be right away).
+				// A typical outcome of a bug here would be use-after-free of
+				// the trace of this client request; we finish it when
+				// returning from here, but the Raft execution also uses it.
+				ctxDone = nil
+				if r.tryAbandon(pendingCmd.idKey) {
+					// TODO(tschottdorf): the command will still execute at
+					// some process, so maybe this should be a structured error
+					// which can be interpreted appropriately upstream.
+					pErr = roachpb.NewError(ctx.Err())
+				} else {
+					log.Warningf("unable to cancel expired Raft command %s", ba)
+				}
+			}
 		}
 	} else {
 		pErr = roachpb.NewError(err)
 	}
 	return br, pErr
+}
+
+// tryAbandon attempts to remove a pending command from the internal commands
+// map. This is possible until execution of the command at the local replica
+// has already begun, in which case false is returned and the client needs to
+// continue waiting for successful execution.
+func (r *Replica) tryAbandon(idKey cmdIDKey) bool {
+	r.mu.Lock()
+	_, ok := r.mu.pendingCmds[idKey]
+	delete(r.mu.pendingCmds, idKey)
+	r.mu.Unlock()
+	return ok
 }
 
 // proposeRaftCommand prepares necessary pending command struct and
@@ -990,8 +1023,9 @@ func (r *Replica) proposeRaftCommand(ctx context.Context, ba roachpb.BatchReques
 	idKeyBuf = encoding.EncodeUint64Ascending(idKeyBuf, uint64(rand.Int63()))
 	idKey := cmdIDKey(idKeyBuf)
 	pendingCmd := &pendingCmd{
-		ctx:  ctx,
-		done: make(chan roachpb.ResponseWithError, 1),
+		ctx:   ctx,
+		idKey: idKey,
+		done:  make(chan roachpb.ResponseWithError, 1),
 		raftCmd: roachpb.RaftCommand{
 			RangeID:       r.RangeID,
 			OriginReplica: *replica,

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -4005,3 +4005,52 @@ func TestGCIncorrectRange(t *testing.T) {
 		t.Errorf("expected value at key %s to no longer exist after GC to correct range, found value %v", key, resVal)
 	}
 }
+
+// TestReplicaCancelRaft checks that it is possible to safely abandon Raft
+// commands via a canceable context.Context.
+func TestReplicaCancelRaft(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	for _, cancelEarly := range []bool{true, false} {
+		func() {
+			// Pick a key unlikely to be used by background processes.
+			key := []byte("acdfg")
+			ctx, cancel := context.WithCancel(context.Background())
+			TestingCommandFilter = func(_ roachpb.StoreID, args roachpb.Request, _ roachpb.Header) error {
+				if !args.Header().Key.Equal(key) {
+					return nil
+				}
+				if cancelEarly {
+					return errors.New("expected client to abandon this request")
+				}
+				cancel()
+				return nil
+			}
+			defer func() { TestingCommandFilter = nil }()
+			tc := testContext{}
+			tc.Start(t)
+			defer tc.Stop()
+			if cancelEarly {
+				cancel()
+			}
+			var ba roachpb.BatchRequest
+			ba.Add(&roachpb.GetRequest{
+				Span: roachpb.Span{Key: key},
+			})
+			br, pErr := tc.rng.addWriteCmd(ctx, ba, nil /* wg */)
+			if pErr == nil {
+				if !cancelEarly {
+					// We cancelled the context while the command was already
+					// being processed, so the client had to wait for successful
+					// execution.
+					return
+				}
+				t.Fatalf("expected an error, but got successful response %+v", br)
+			}
+			// If we cancelled the context early enough, we expect to receive a
+			// corresponding error and not wait for the command.
+			if !testutils.IsPError(pErr, context.Canceled.Error()) {
+				t.Fatalf("unexpected error: %s", pErr)
+			}
+		}()
+	}
+}

